### DISCUSSION TITLE
Retrieve all the versions of a PDF when saved incrementally

### DISCUSSION THREAD

**Discussion by 3051360** [pymupdf/PyMuPDF#3096] [Q&A]
I've observed that PyMuPDF provides functionality to determine [the number of versions present in a document](https://pymupdf.readthedocs.io/en/latest/document.html#Document.version_count).

Currently, I'm utilizing [pdfresurrect](https://github.com/enferex/pdfresurrect), a C tool/library designed to "**extract older 'hidden' versions of a PDF from the current PDF.**"

While my larger application is built using PyMuPDF, I have to call the compiled C code through Python for version retrieval, which is not an ideal way of building applications. 

I propose that enhancing the capability to retrieve all PDF version files directly within PyMuPDF would be beneficial.

**Comment by cipri-tom** [comment]
Here's how I do it:

```python

        pattern = re.compile(b'[^t]xref.*?EOF', flags=re.M+re.S)

        # `inc_nb` means the number of incremental update contained in the binary
        matches = list(pattern.finditer(doc.data))
        inc_nb = len(matches)
        if not inc_nb > 1:
            return result

        # here we loop to separate each incremental update in
        # the binary to retrieve each version of the file
        # We don't take into account the current version
        for sub in matches[:-1]:
            stream = doc.data[:sub.span()[1]] # binary code of a version
            if not stream:
                continue
            previous = Doc(stream).as_fitz() # is one of the previous version of the file

```

**Reply by 3051360** [reply]
@cipri-tom Thanks for sharing the approach. I get the essence of what you are trying to do. I think you have provided a part of the code which at the end could result in getting all the versions of an incrementally saved PDF file.

I attempted to complete the overall code (with some dummy code in between) but obviously did not succeed. 


```python

import fitz
import re
import os

def extract_versions(input_pdf_path, output_folder):
    doc = fitz.open(input_pdf_path)

    pattern = re.compile(b'[^t]xref.*?EOF', flags=re.M + re.S)

    matches = list(pattern.finditer(doc.get_raw_data()))

    if len(matches) <= 1:
        print("Only one version found. No incremental updates.")
        return

    os.makedirs(output_folder, exist_ok=True)

    for i, sub in enumerate(matches[:-1]):
        start_offset, end_offset = sub.span()
        version_data = doc.get_raw_data()[start_offset:end_offset]

        version_doc = fitz.open()

        version_doc.insert_pdf(fitz.open(fitz.Stream(fitz.SOString, version_data)))

        output_path = os.path.join(output_folder, f"version_{i + 1}.pdf")
        version_doc.save(output_path)
        print(f"Version {i + 1} saved to: {output_path}")

        version_doc.close()

    doc.close()


```

I am not sure how to proceed further but maybe someone looking for a similar feature could find this code useful.

As @JorjMcKie mentioned in the later thread, I will drop a request to people at MuPDF for this feature.

**Reply by cipri-tom** [reply]
the problem here is that you get `version_data` from `start_offset` to `end_offset`. Instead, you should always get it from the start, until the `end_offset`. This is why it is "incremental", new data is appended, working in addition to existing data.

**Comment by JorjMcKie** [comment]
This feature would have to be implemented by our base library MuPDF to make any sense.
Currently, you can find out the number of versions of a PDF: `doc.version_count` only.

Please talk to the MuPDF colleagues on e.g. their [Dicscord channel](https://discord.gg/9nbEAeUn) about options they may see.

**Reply by 3051360** [reply]
@JorjMcKie  Thank you for your response, I will reach out to people at MuPDF to see if they could point me to something.

**Comment by mlissner** [comment]
I'm working on adding document versioning support to MuPDF, assuming they welcome the contributions. If I do that, will folks here be interested in adding it to PyMuPDF (@JorjMcKie, that still you? Been a while!) I could probably do a PR for that too after MuPDF is merged and released, but I'd certainly welcome help on this if possible and I can get it into MuPDF.

**Update**: Here's the PR: https://github.com/ArtifexSoftware/mupdf/pull/88

**Reply by JorjMcKie** [reply]
Hi @mlissner , yes I am still on board -  now together with other maintainers. 
We will continue to reflect MuPDF features in pymupdf,  and this certainly also applies to what you are working on!
We currently can just return the count of the versions, there is no way to show a file in a non-current state. 
We do however support mupdf's logging feature.  This includes the option to save a pdf incrementally with each commit representing a pdf incremental update. If the corresponding log file is available,  the commits/versions can be "undone"/"redone" ... thus representing a feature similar to your intentions.
